scripts/build/build/__init__.py

```diff
@@ -0,0 +1,151 @@
+APPLICATIONS = [x.ArgName for x in Application]
+
+
+class BuildSteps(Enum):
+  GENERATED = auto()
+
+
+class Context:
+  """Represents a grouped list of platform/board/app builders to use
+
+       to generate make/ninja instructions and to compile.
+    """
+
+  def __init__(self, repository_path, output_prefix):
+    self.builders = []
+    self.repository_path = repository_path
+    self.output_prefix = output_prefix
+    self.completed_steps = set()
+
+  def SetupBuilders(self, platforms: Sequence[Platform],
+                    boards: Sequence[Board],
+                    applications: Sequence[Application]):
+    """Configures internal builders for the given platform/board/app combionation.
+
+        Handles smart default selection, so that users only need to specify
+        part of platform/board/application information and the method tries
+        to automatically deduce the rest of the arguments.
+        """
+    if not platforms and not boards:
+      if applications:
+        platforms = set().union(*[
+            TargetRelations.PlatformsForApplication(app) for app in applications
+        ])
+      else:
+        # when nothing is specified, start with a default host build
+        platforms = [Platform.LINUX]
+
+    # at this point, at least one of 'platforms' or 'boards' is non-empty
+    if not boards:
+      boards = set().union(*[
+          TargetRelations.BoardsForPlatform(platform) for platform in platforms
+      ])
+    elif not platforms:
+      platforms = set().union(
+          *[TargetRelations.PlatformsForBoard(board) for board in boards])
+
+    if not applications:
+      applications = set().union(*[
+          TargetRelations.ApplicationsForPlatform(platform)
+          for platform in platforms
+      ])
+
+    platforms = set(platforms)
+    boards = set(boards)
+    applications = set(applications)
+
+    logging.info('Platforms being built: %s', CommaSeparate(platforms))
+    logging.info('Boards being built: %s', CommaSeparate(boards))
+    logging.info('Applications being built: %s', CommaSeparate(applications))
+
+    # Sanity check: ensure all input arguments generate at least an output
+    platforms_with_builders = set()
+    boards_with_builders = set()
+    applications_with_builders = set()
+
+    factory = BuilderFactory(self.repository_path, self.output_prefix)
+
+    for platform in platforms:
+      for board in boards:
+        for application in applications:
+          builder = factory.Create(platform, board, application)
+          if not builder:
+            logging.debug('Builder not supported for tuple %s/%s/%s', platform,
+                          board, application)
+            continue
+
+          self.builders.append(builder)
+          platforms_with_builders.add(platform)
+          boards_with_builders.add(board)
+          applications_with_builders.add(application)
+
+    if platforms != platforms_with_builders:
+      logging.warn('Platforms without build output: %s',
```

Comment:
Why is this not fatal?

Reply:
I assumed that people would just list a bunch of things regardless if compatible or not. The build output is already named as 'paltform-board-app' combination and the script generally accepts fuzzyness: '--platform all --app lock' will be interpreted as "build the lock app on all supported platforms" rather than "build the lock on all platforms" (key being "SUPPORTED")

since many apps are not actually supported for some example apps, I believe fuzzy is better than failing. With failing, I am guaranteed to error out on any "--platform all" or we have to special case.

I am ok with special casing (i.e. if non 'all' was in the args, complain if no match) however I would rather do that separately - it is a more complex change and I believe it needs some thought on what to provide (maybe a "--strict" argument or "--nofuzzy" or something).

Reply:
> I assumed that people would just list a bunch of things regardless if compatible or not. The build output is already named as 'paltform-board-app' combination and the script generally accepts fuzzyness: '--platform all --app lock' will be interpreted as "build the lock app on all supported platforms" rather than "build the lock on all platforms" (key being "SUPPORTED")
> 
> since many apps are not actually supported for some example apps, I believe fuzzy is better than failing. With failing, I am guaranteed to error out on any "--platform all" or we have to special case.
> 
> I am ok with special casing (i.e. if non 'all' was in the args, complain if no match) however I would rather do that separately - it is a more complex change and I believe it needs some thought on what to provide (maybe a "--strict" argument or "--nofuzzy" or something).

I have no objection to the fuzzy behavior for something that is already a wildcard like "all" but if it is possible to ask for a specific combination then I think it is quite a bad outcome if the script can succeed without building that combination.